win32ss/gdi/ntgdi/freetype.c

```diff
@@ -715,6 +711,83 @@ InitFontSupport(VOID)
     return TRUE;
 }
 
+static VOID
+IntFreeFontCacheList(PLIST_ENTRY pHead)
+{
+    PLIST_ENTRY pEntry, pNextEntry;
+    PFONT_CACHE_ENTRY pFontCache;
+
+    ASSERT_FREETYPE_LOCK_HELD();
+
+    for (pEntry = pHead->Flink; pEntry != pHead; pEntry = pNextEntry)
+    {
+        pNextEntry = pEntry->Flink;
+        pFontCache = CONTAINING_RECORD(pEntry, FONT_CACHE_ENTRY, ListEntry);
+        RemoveCachedEntry(pFontCache);
+    }
+
+    pHead->Flink = pHead;
+}
+
+static VOID
+IntFreeFontSubstList(PLIST_ENTRY pHead)
+{
+    PLIST_ENTRY pEntry, pNextEntry;
+    PFONTSUBST_ENTRY pSubstEntry;
+
+    ASSERT_FREETYPE_LOCK_HELD();
+
+    for (pEntry = pHead->Flink; pEntry != pHead; pEntry = pNextEntry)
+    {
+        pNextEntry = pEntry->Flink;
+        pSubstEntry = CONTAINING_RECORD(pEntry, FONTSUBST_ENTRY, ListEntry);
+        ExFreePoolWithTag(pSubstEntry, TAG_FONT);
+    }
+
+    pHead->Flink = pHead;
+}
+
+static VOID
+IntFreeFontList(PLIST_ENTRY pHead)
+{
+    PLIST_ENTRY pEntry, pNextEntry;
+    PFONT_ENTRY pFontEntry;
+
+    ASSERT_FREETYPE_LOCK_HELD();
+
+    for (pEntry = pHead->Flink; pEntry != pHead; pEntry = pNextEntry)
+    {
+        pNextEntry = pEntry->Flink;
+        pFontEntry = CONTAINING_RECORD(pEntry, FONT_ENTRY, ListEntry);
+        CleanupFontEntry(pFontEntry);
+    }
+
+    pHead->Flink = pHead;
+}
+
+VOID FASTCALL FreeFontSupport(VOID)
+{
+    IntLockGlobalFonts();
+
+    IntFreeFontCacheList(&g_FontCacheListHead);
+    IntFreeFontSubstList(&g_FontSubstListHead);
+    IntFreeFontList(&g_FontListHead);
```

Comment:
I'm not fully convinced you need to have three separate helpers just to use them only here.
Their implementation could directly be moved within there.

Reply:
Agreed. They can be moved inside this function as well.

Reply:
Fixed.